scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,370 @@
+	u64 now = bpf_ktime_get_ns();
+
+	ts = bpf_task_storage_get(&task_storage_map, p, 0, 0);
+	if (!ts)
+		return;
+	if (update_start)
+		ts->start_ns = now;
+	if (update_stop)
+		ts->stop_ns = now;
+}
+
+/* Task @p starts to run on a CPU */
+void BPF_STRUCT_OPS(rustlite_running, struct task_struct *p)
+{
+	bpf_printk("start: pid=%d (%s)", p->pid, p->comm);
+	update_task_time(p, true, false);
+}
+
+/* Task @p releases a CPU */
+void BPF_STRUCT_OPS(rustlite_stopping, struct task_struct *p, bool runnable)
+{
+	bpf_printk("stop: pid=%d (%s)", p->pid, p->comm);
+	update_task_time(p, false, true);
+}
+
+/*
+ * Return the time slice (in ns) used by the task during the previous execution
+ * slot (up to a maximum of slice_ns = full time slice used).
+ */
+static u64 task_slice_used(struct task_struct *p)
+{
+	struct task_storage *ts;
+	u64 time_ns;
+
+	ts = bpf_task_storage_get(&task_storage_map, p, 0, 0);
+	if (!ts)
+		return 0;
+	time_ns = time_diff(ts->stop_ns, ts->start_ns);
+
+	return MIN(time_ns, slice_ns);
+}
+
+/*
+ * Return task's vtime, evaluated as following:
+ *
+ *    current time + used task slice (scaled based on task's priority)
+ */
+static u64 task_vtime(struct task_struct *p)
+{
+	return bpf_ktime_get_ns() + task_slice_used(p) * 100 / p->scx.weight;
+}
+
+/*
+ * Task @p starts to be controlled by the sched-ext scheduling class for the
+ * first time.
+ */
+void BPF_STRUCT_OPS(rustlite_enable, struct task_struct *p,
+		    struct scx_enable_args *args)
+{
+	bpf_printk("new: pid=%d (%s)", p->pid, p->comm);
+	update_task_time(p, true, true);
+}
+
+/*
+ * Initilize local storage for a new task.
+ *
+ * We can sleep in this function and the allocation of the local task storage
+ * will automatically use GFP_KERNEL.
+ */
+s32 BPF_STRUCT_OPS(rustlite_prep_enable, struct task_struct *p,
+		   struct scx_enable_args *args)
+{
+	if (bpf_task_storage_get(&task_storage_map, p, 0,
+				 BPF_LOCAL_STORAGE_GET_F_CREATE))
+		return 0;
+	else
+		return -ENOMEM;
+}
+
+/* Dispatch a task on the local per-CPU FIFO */
+static inline void dispatch_task_local(struct task_struct *p, u64 enq_flags)
+{
+	scx_bpf_dispatch(p, SCX_DSQ_LOCAL, slice_ns, enq_flags);
+	__sync_fetch_and_add(&nr_kernel_dispatches, 1);
+}
+
+/* Dispatch a task on the global FIFO */
+static inline void dispatch_task_global(struct task_struct *p, u64 enq_flags)
+{
+	scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, enq_flags);
+	__sync_fetch_and_add(&nr_user_dispatches, 1);
+}
+
+/* Task @p becomes ready to run */
+void BPF_STRUCT_OPS(rustlite_enqueue, struct task_struct *p, u64 enq_flags)
+{
+        struct task_ctx task;
+
+	/*
+	 * Scheduler is dispatched directly in .dispatch() when needed, so
+	 * we can skip it here.
+	 */
+	if (is_usersched_task(p))
+		return;
+
+        /*
+	 * Dispatch per-cpu kthreads on the local FIFO directly from the
+	 * kernel.
+         */
+	if (is_kthread(p) && p->nr_cpus_allowed == 1) {
+		bpf_printk("dispatch (local): pid=%d", p->pid);
```

Comment:
Yes, we can bring all the dispatch bpf_printk into `dispatch_task_local/global()`, maybe also add a vtime argument for global, so that we can bring in also the bpf_printk in `rustlite_dispatch()`?